app/controllers/application_controller.rb

```diff
@@ -1,4 +1,31 @@
 class ApplicationController < ActionController::Base
   include AllowBrowser, Authentication, Authorization, SetCurrentRequest, SetPlatform, TrackedRoomVisit, VersionHeaders
   include Turbo::Streams::Broadcasts, Turbo::Streams::StreamName
+
+  before_action :forbid_banned_user
+
+  private
+    def forbid_banned_user
+      return unless Current.user&.banned?
+      return if banned_user_allowed_action?
+
+      # GG: flash seems is unable to be shown, so we do not put it
+      # flash.now[:alert] = "Your account has been banned."
+
+      if turbo_request?
+        render turbo_stream: turbo_stream.update("main-content", partial: "application/banned_notice"), status: :forbidden
+      elsif request.format.html?
+        render "application/banned", status: :forbidden
+      else
+        head :forbidden
+      end
+    end
+
+    def banned_user_allowed_action?
+      controller_path == "sessions" || controller_path.start_with?("sessions/")
+    end
```

Comment:
Maybe it would be nicer if this was an action like the session/auth actions are?
Then, at the top of the SessionsController, you could add something like `allow_access_while_banned only: :destroy`.